### ISSUE TITLE
Feature Tests

### ISSUE THREAD

**Issue by samcaulfield** [samcaulfield/pml-studio#25]
We require a set of tests to verify that each feature in the backlog probably works.

We should begin with a broad coverage of the feature set. The initial tests should look for major faults in functionality. We can then add more detailed test cases once the full feature set is covered by some tests.

**Comment by samcaulfield** [comment]
I will update the README to include testing instructions for the new features.

The test suite will need to automate the tests listed in the README.

@gallagts what is the purpose of the Test directory?

**Comment by samcaulfield** [comment]
Important note: John said he will waiver the requirement of testing the graphical editor.

**Comment by samcaulfield** [comment]
I've taken a look at the uploaded selenium Python script and have some observations:

- The uploaded file has no file extension. The file extension should be `.py`.

- We should choose a consistent name for the test project, perhaps `seleniumTest` instead of `r9` which could interfere with already installed projects.

- We will need to add the `python-selenium` Ubuntu package as a dependency.

- We will need to add the command `sudo pip -U selenium` to the installation process to update selenium to a version that works with the latest Firefox browser.

- The uploaded python script fails. Lines `22` and `23` need to be switched for a start. I'm curious how the generated script has this kind of error. There are more errors but I feel like I should hand back to you at this stage until you can generate a script that passes.

- The script opens Firefox and performs the UI interactions. This might not be suitable for a pre-install test. We may have to add a new `test` target to the Makefile for these tests.

- The test names should probably match the feature name on the product backlog. So `registerSignInTest` should be `authenticationTest`. This is so we don't get a headache later on trying to figure out if every feature is tested.

**Comment by seanlth** [comment]
Good plan.

**Comment by Gulnur1** [comment]
When testing the Signing in with a Google account, the Sign In button is not actually a button but a div, I didn't find a way to click on a div to test that part of the authentication.

**Comment by samcaulfield** [comment]
Perhaps the information in this [link](http://selenium-python.readthedocs.org/locating-elements.html?highlight=By) could help you programmatically locate and click on the div.

**Comment by Gulnur1** [comment]
Hey Sam, do you know how to drop the table in the database?

**Comment by Gulnur1** [comment]
@sharonluk , @gallagts , I am doing syntax analysis feature test, what features are you testing atm?

**Comment by samcaulfield** [comment]
I believe the SQL command for dropping a table is `drop table <tableName>`. In PHP you could do this with `$db->query('drop table <tableName>');` after setting up a `db` object which is done in a few of the PHP scripts in the repo.

**Comment by Gulnur1** [comment]
What is the name of the table that contains usernames and passwords that were registered in the db ? Is there any password to connect to the db? Or specific parameters(specific port or something) that I should know when connecting ?

**Comment by Gulnur1** [comment]
I had done syntax analysis & register, sign in , sign out, but not the thirdparty registration.

**Comment by samcaulfield** [comment]
[This](https://github.com/samcaulfield/pml-studio/blob/master/php/register.php) script should have what you need. The users database is called _user.db_ and contains a table called _user_. The user table contains email and password pairs. It's not passworded. It's an SQLite database if that helps.

**Comment by Gulnur1** [comment]
Are we only meant to test features for release 2?

**Comment by samcaulfield** [comment]
I think the expectation is that we have all features tested (except boxes & arrows)

**Comment by sharonluk** [comment]
Hey , I managed to get test file working for users being able to specify the name to be saved.

**Comment by Gulnur1** [comment]
@sharonluk 
1. Why did you delete the code you submitted?  
2. Can you specify the project name as seleniumTest for this and future tests because that is the projectName we will use for testing.

**Comment by sharonluk** [comment]
yes will do, it was running on my computer saying the test has passed, everything "ok" but when tested on a different desktop, there was error. I will have to modify it again before i commit.

**Comment by Gulnur1** [comment]
you need to install the project name on the desktop you are running it as seleniumTest and then change in your code the project name to that as well

**Comment by sharonluk** [comment]
I will make those changes.

**Comment by Gulnur1** [comment]
@samcaulfield , is there any way to get the project name from where the user installs it? Where is it and how is it saved? 
For the tests, we need to be able to get the project name for every test.

**Comment by samcaulfield** [comment]
The project name is the name of the folder in /var/www/html that the project files reside in. However we probably don't want to install the test scripts along with the project files during installation as the tests aren't part of the functionality.

**Comment by Gulnur1** [comment]
Then, maybe I will put an installation command with the project name specified in the test.sh so when it is run it can install and run all of the tests.

**Comment by samcaulfield** [comment]
I think it would actually be better if we didn't have a specific project for testing, and allows the user to specify which instance to test. A `test` target could be added to the Makefile to achieve this. The command would be `make test ProjectName=foobar`

**Comment by Gulnur1** [comment]
okay, that's a good idea.

**Comment by Gulnur1** [comment]
@sharonluk @gallagts , can you instead of 
**driver.get(self.base_url + "/selenuimTest/")**  in your tests  put these lines below: 
`f = open('property')`
`project_name = f.read()`
`f.close()`
`driver.get(self.base_url + "/" + project_name + "/")`

**Comment by sharonluk** [comment]
hey Gulnur, when i test your registerSignIn file, property only contains "whatever". 
Do we need to manually put it in or is there some way property is meant to contain a name?

**Comment by Gulnur1** [comment]
@sharonluk  , 
1.  make install ProjectName=rr9, or rr10, or whatever you want to call the project. 
2. make test ProjectName=rr9, or rr10, or whatever you called the project in the first step. 
The first test will fail, the rest will be OK. I will fix the first test later. 

The property file just saves the name of the project and deletes every time you install it a different name.

**Comment by Gulnur1** [comment]
@sharonluk  do not modify the property file

**Comment by samcaulfield** [comment]
Several tests are failing for me on a clean install. My system is Ubuntu 14.04 with Firefox 45.0.

Commands used:
`make install ProjectName=editor`
`make test ProjectName=editor`

Log file attached.

[log.txt](https://github.com/samcaulfield/pml-studio/files/209036/log.txt)

**Comment by Gulnur1** [comment]
@samcaulfield can you rerun tests now? Probably your system is faster than mine, I put time.sleep before the lines that were failing.

**Comment by samcaulfield** [comment]
All the tests but one pass for me now. Here is the output from the failed test:
```
Now testing for keyword completion
The test has passed for requires keyword
E
======================================================================
ERROR: test_keyword_completion (__main__.KeywordCompletion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "keywordAndResourceCompletion.py", line 38, in test_keyword_completion
    raise Exception ("Keyword test for %s failed!!!!!" % dic[key])
Exception: Keyword test for action failed!!!!!

----------------------------------------------------------------------
Ran 1 test in 6.518s

FAILED (errors=1)
make: *** [test] Error 1
```

If the tests are sensitive to the machine's speed then I think it would be a recipe for disaster to hard code sleeps into the script. Perhaps [implicit waits](http://selenium-python.readthedocs.org/waits.html#explicit-waits) would fix this robustly.

**Comment by Gulnur1** [comment]
I already tried waits, they did nothing.

**Comment by Gulnur1** [comment]
I will fix the last test with implicit waits and we'll se if it works then

**Comment by Gulnur1** [comment]
@samcaulfield Can you try now? I can't check on mine because it works for me and it's only meant to wait if the element id is not immediately available, it should wait in your system.

**Comment by samcaulfield** [comment]
It works for me now. I'll test again on an even faster machine tomorrow to see if it holds up.

**Comment by Gulnur1** [comment]
Okay, I will change the sleeps to implicit waits then.

**Comment by samcaulfield** [comment]
Several tests fail for me on a very powerful Ubuntu 14.04 machine running Firefox 44.0.2. Log attached.

This seems to be caused by the same issue as before. Unfortunately implicit waits are not a robust solution to this issue. Are you sure explicit waits didn't work for you?

[log.txt](https://github.com/samcaulfield/pml-studio/files/210631/log.txt)

**Comment by Gulnur1** [comment]
I tried it before for the tests failing on my machine, I will try to put them back again, can you test it on your machine again when I'm done? If that doesn't work, I will have to use the sleeps.

**Comment by Gulnur1** [comment]
@samcaulfield , can you try to run the registerSignInSignOut.py only on the same conditions that the tests failed when running ? I added explicit wait before the robust instruction.

**Comment by samcaulfield** [comment]
Test still does not pass. I am investigating further.

**Comment by Gulnur1** [comment]
I did all the tests except for : 

1. `Jump to next/prev error`, because at the moment it does not allow moving to next/prev errors, it only gives the first error;
2. `Save to/open from disk` - @gallagts is doing those;
3. I was not able to produce tests for `Syntax highlighting` and `Third party authentication`, but will try again. 
4. From the `Code editor` feature everything is done except for syntax highlighting test. 

Also, I will look into waits, I can put a Selenium `pause` command and try with that. It seems to jump to the next command too fast before the previous instruction is done if the machine is very powerful.

**Comment by samcaulfield** [comment]
Gulnur can you confirm that all tests are passing for you? If so, what system are you testing on? I have run the tests in a VM on my laptop and multiple tests still fail.

**Comment by Gulnur1** [comment]
Yes, everything is running on my machine, Ubuntu 14.04 VM.  Firefox 44.0.2.

**Comment by Gulnur1** [comment]
Sam, which tests are failing for you? 
Did anyone else run the tests on their machines?

**Comment by samcaulfield** [comment]
On the latest version (e0f4359cb76b6917956c8998e5e99b894d22e3c6), Ubuntu 14.04 and Firefox 45.0 I running the tests on the clean install yields the following output: [log.txt](https://github.com/samcaulfield/pml-studio/files/211446/log.txt)

It looks like the tests that are failing for me are registerSignInSignOut, all the syntax analysis tests, saveRetrieveServer.

**Comment by Gulnur1** [comment]
Sam, now the tests should be working for you except for the saveRetrieveServer.

**Comment by Gulnur1** [comment]
There is one error when Registering, it does not automatically sign in when registering, it says `Undefined` instead of the email address that should be displayed as signed in. It was working before yesterday when I updated Firefox or some other external changes have been made. 
How is the save to/open from disk test going, Troy?

**Comment by Gulnur1** [comment]
The method getCookie in script.js returns undefined, which is why when registering in testing it returns undefined for the email address. 
I don't know how to deal with that, which is why now the test for save and retrieve from server does not work.

**Comment by samcaulfield** [comment]
I can't reproduce that behaviour by following the same actions manually. Is this an issue on the Selenium side?

**Comment by Gulnur1** [comment]
Yes, it is only for the testing that it does not work.

**Comment by samcaulfield** [comment]
What shall we do about test_save_retrieve_server? Selenium seems to be having trouble with cookies resulting in tests that have to check `isLoggedIn` failing. I think it should be commented out for the release.

It always fails so won't tell us anything new when it fails.

**Comment by Gulnur1** [comment]
Yes, I agree. We should exclude the one failing as it may also give an impression that the feature is not working properly. I will take it off there then.

**Comment by Gulnur1** [comment]
Everything else working for you, Sam?

**Comment by samcaulfield** [comment]
Yes, all the other tests pass for me.
